{"cells":[{"cell_type":"code","source":["import numpy as np\n","import matplotlib.pyplot as plt\n","import tensorflow as tf\n","from tensorflow.keras.models import Model\n","from tensorflow.keras.layers import Input, Dense\n","from tensorflow.keras.callbacks import EarlyStopping"],"metadata":{"id":"T00pNQy2kzUD","executionInfo":{"status":"ok","timestamp":1720965520600,"user_tz":180,"elapsed":408,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":12,"outputs":[]},{"cell_type":"code","source":["# Função de normalização dos dados\n","def normalize_data(data):\n","    data_min = np.min(data, axis=0)\n","    data_max = np.max(data, axis=0)\n","    data = (data - data_min) / (data_max - data_min)\n","    return data"],"metadata":{"id":"i4qELLs6xu9f","executionInfo":{"status":"ok","timestamp":1720965301899,"user_tz":180,"elapsed":33,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":2,"outputs":[]},{"cell_type":"code","source":["# Definição da arquitetura do autoencoder\n","def build_autoencoder(input_dim, encoding_dim):\n","    input_layer = Input(shape=(input_dim,))\n","    encoded = Dense(encoding_dim, activation='relu')(input_layer)\n","    decoded = Dense(input_dim, activation='softmax')(encoded)\n","\n","    autoencoder = Model(inputs=input_layer, outputs=decoded)\n","\n","    # Define o encoder e decoder separadamente\n","    encoder = Model(inputs=input_layer, outputs=encoded)\n","    encoded_input = Input(shape=(encoding_dim,))\n","    decoder_layer = autoencoder.layers[-1]\n","    decoder = Model(inputs=encoded_input, outputs=decoder_layer(encoded_input))\n","\n","    autoencoder.compile(optimizer='adam', loss='sparse_categorical_crossentropy')\n","\n","    # Adiciona propriedades ao autoencoder\n","    autoencoder.encoder = encoder\n","    autoencoder.decoder = decoder\n","\n","    return autoencoder\n"],"metadata":{"id":"aLaRLrhFx1vQ","executionInfo":{"status":"ok","timestamp":1720965533758,"user_tz":180,"elapsed":323,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":13,"outputs":[]},{"cell_type":"code","source":["# Função de treinamento com Early Stopping\n","def train_autoencoder_with_early_stopping(autoencoder, data, labels, val_data, val_labels, epochs=100, patience=10):\n","    early_stopping = EarlyStopping(monitor='val_loss', patience=patience, verbose=1, restore_best_weights=True)\n","    history = autoencoder.fit(data, labels, epochs=epochs, batch_size=32, validation_data=(val_data, val_labels), callbacks=[early_stopping])\n","    return history"],"metadata":{"id":"AyurgOUfwW7D","executionInfo":{"status":"ok","timestamp":1720965545584,"user_tz":180,"elapsed":313,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":14,"outputs":[]},{"cell_type":"code","source":["# Função de plotagem da evolução do treinamento\n","def plot_training_history(history):\n","    plt.figure(figsize=(10, 6))\n","    plt.plot(history.history['loss'], label='Training Loss')\n","    plt.plot(history.history['val_loss'], label='Validation Loss')\n","    plt.xlabel('Epochs')\n","    plt.ylabel('Loss')\n","    plt.title('Training and Validation Loss')\n","    plt.legend()\n","    plt.grid(True)\n","    plt.show()\n"],"metadata":{"id":"tvLHPZpHwdes","executionInfo":{"status":"ok","timestamp":1720965609526,"user_tz":180,"elapsed":329,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":16,"outputs":[]},{"cell_type":"code","source":["# Função de teste para calcular a SER\n","def test_autoencoder_with_errors(autoencoder, data, labels, EbNodB_range, R):\n","    ber = np.zeros(len(EbNodB_range))\n","    num_errors = []\n","\n","    for n in range(len(EbNodB_range)):\n","        EbNo = 10.0 ** (EbNodB_range[n] / 10.0)\n","        noise_std = np.sqrt(1 / (2 * R * EbNo))\n","        noise = np.random.normal(0, noise_std, data.shape)\n","\n","        encoded_signal = autoencoder.encoder.predict(data)\n","        channel_signal = encoded_signal + noise\n","        decoded_signal = autoencoder.decoder.predict(channel_signal)\n","\n","        pred_output = np.argmax(decoded_signal, axis=1)\n","        errors = np.sum(pred_output != labels)\n","        ber[n] = errors / len(labels)\n","        num_errors.append(errors)\n","\n","        print(f'SNR: {EbNodB_range[n]} BLER: {ber[n]} no_errors: {errors}')\n","\n","    return ber, num_errors"],"metadata":{"id":"ZMzGo-ewwfM3","executionInfo":{"status":"ok","timestamp":1720965638942,"user_tz":180,"elapsed":310,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":17,"outputs":[]},{"cell_type":"code","source":["# Função para visualizar o resultado do encoder\n","def plot_encoded_space(encoder, data, labels):\n","    encoded_data = encoder.predict(data)\n","    plt.figure(figsize=(10, 6))\n","    plt.scatter(encoded_data[:, 0], encoded_data[:, 1], c=labels, cmap='viridis')\n","    plt.colorbar()\n","    plt.xlabel('Encoded Dimension 1')\n","    plt.ylabel('Encoded Dimension 2')\n","    plt.title('Encoded Space')\n","    plt.grid(True)\n","    plt.show()"],"metadata":{"id":"kFKFYi58wkcS","executionInfo":{"status":"ok","timestamp":1720965688269,"user_tz":180,"elapsed":307,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":18,"outputs":[]},{"cell_type":"code","source":["# Função de definição dos parâmetros de treino\n","def prepare_data(n, k):\n","    data_size = 1000\n","    data = np.random.randint(0, 2, size=(data_size, n))\n","    labels = np.packbits(data, axis=-1)\n","    labels = np.squeeze(labels)\n","\n","    val_data = np.random.randint(0, 2, size=(200, n))\n","    val_labels = np.packbits(val_data, axis=-1)\n","    val_labels = np.squeeze(val_labels)\n","\n","    data = normalize_data(data)\n","    val_data = normalize_data(val_data)\n","\n","    return data, labels, val_data, val_labels"],"metadata":{"id":"ogQVpczRzlSH","executionInfo":{"status":"ok","timestamp":1720965724074,"user_tz":180,"elapsed":319,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":19,"outputs":[]},{"cell_type":"code","source":["# Função de plotagem do número de erros por SNR\n","def plot_errors_by_snr(num_errors, EbNodB_range):\n","    plt.figure(figsize=(10, 6))\n","    plt.plot(EbNodB_range, num_errors, marker='o', label='Number of Errors')\n","    plt.xlabel('SNR (dB)')\n","    plt.ylabel('Number of Errors')\n","    plt.title('Number of Errors vs SNR')\n","    plt.legend()\n","    plt.grid(True)\n","    plt.show()"],"metadata":{"id":"IWOzL_kSzqzu","executionInfo":{"status":"ok","timestamp":1720965738839,"user_tz":180,"elapsed":308,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":20,"outputs":[]},{"cell_type":"code","source":["# Parâmetros gerais\n","EbNodB_range = np.linspace(0, 10, 11)\n","R = 1"],"metadata":{"id":"gPuL6xZrzu_F","executionInfo":{"status":"ok","timestamp":1720965750742,"user_tz":180,"elapsed":305,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}}},"execution_count":21,"outputs":[]},{"cell_type":"code","source":["# Exercício a: n=1, k=2, modulação PSK\n","data_a, labels_a, val_data_a, val_labels_a = prepare_data(1, 2)\n","autoencoder_a = build_autoencoder(1, 2)\n","history_a = train_autoencoder_with_early_stopping(autoencoder_a, data_a, labels_a, val_data_a, val_labels_a)\n","plot_training_history(history_a)\n","ber_a, num_errors_a = test_autoencoder_with_errors(autoencoder_a, data_a, labels_a, EbNodB_range, R)\n","plot_encoded_space(autoencoder_a.encoder, data_a, labels_a)\n","plot_errors_by_snr(num_errors_a, EbNodB_range)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"id":"HH2sbSkczz_0","executionInfo":{"status":"error","timestamp":1720965765175,"user_tz":180,"elapsed":1346,"user":{"displayName":"Gabriel Francisco","userId":"09939884952713407859"}},"outputId":"a5800631-2cfd-4e4a-d461-85d4c6abf886"},"execution_count":22,"outputs":[{"output_type":"stream","name":"stdout","text":["Epoch 1/100\n"]},{"output_type":"error","ename":"InvalidArgumentError","evalue":"Graph execution error:\n\nDetected at node sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits defined at (most recent call last):\n  File \"/usr/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n\n  File \"/usr/lib/python3.10/runpy.py\", line 86, in _run_code\n\n  File \"/usr/local/lib/python3.10/dist-packages/colab_kernel_launcher.py\", line 37, in <module>\n\n  File \"/usr/local/lib/python3.10/dist-packages/traitlets/config/application.py\", line 992, in launch_instance\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelapp.py\", line 619, in start\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/platform/asyncio.py\", line 195, in start\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n\n  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 685, in <lambda>\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 738, in _run_callback\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 825, in inner\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 786, in run\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 361, in process_one\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 261, in dispatch_shell\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 539, in execute_request\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py\", line 302, in do_execute\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/zmqshell.py\", line 539, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2975, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3030, in _run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3257, in run_cell_async\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n\n  File \"<ipython-input-22-ddbb2e7932da>\", line 4, in <cell line: 4>\n\n  File \"<ipython-input-14-e97dc6108395>\", line 4, in train_autoencoder_with_early_stopping\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 65, in error_handler\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1807, in fit\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1401, in train_function\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1384, in step_function\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1373, in run_step\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1151, in train_step\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1209, in compute_loss\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 277, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 143, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 270, in call\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 2454, in sparse_categorical_crossentropy\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend.py\", line 5775, in sparse_categorical_crossentropy\n\nReceived a label value of 128 which is outside the valid range of [0, 1).  Label values: 0 128 0 128 128 0 0 0 128 0 128 128 0 128 128 0 128 0 0 128 128 128 128 0 0 128 0 0 128 128 0 0\n\t [[{{node sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits}}]] [Op:__inference_train_function_838]","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mInvalidArgumentError\u001b[0m                      Traceback (most recent call last)","\u001b[0;32m<ipython-input-22-ddbb2e7932da>\u001b[0m in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mdata_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_data_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_labels_a\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprepare_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mautoencoder_a\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbuild_autoencoder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mhistory_a\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtrain_autoencoder_with_early_stopping\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mautoencoder_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_data_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_labels_a\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mplot_training_history\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhistory_a\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mber_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_errors_a\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtest_autoencoder_with_errors\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mautoencoder_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels_a\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mEbNodB_range\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mR\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m<ipython-input-14-e97dc6108395>\u001b[0m in \u001b[0;36mtrain_autoencoder_with_early_stopping\u001b[0;34m(autoencoder, data, labels, val_data, val_labels, epochs, patience)\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mtrain_autoencoder_with_early_stopping\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mautoencoder\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpatience\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mearly_stopping\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mEarlyStopping\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmonitor\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'val_loss'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpatience\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpatience\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrestore_best_weights\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mhistory\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mautoencoder\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mepochs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m32\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalidation_data\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_labels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallbacks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mearly_stopping\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mhistory\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     68\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     69\u001b[0m             \u001b[0;31m# `tf.debugging.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 70\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     71\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tensorflow/python/eager/execute.py\u001b[0m in \u001b[0;36mquick_execute\u001b[0;34m(op_name, num_outputs, inputs, attrs, ctx, name)\u001b[0m\n\u001b[1;32m     51\u001b[0m   \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m     \u001b[0mctx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mensure_initialized\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,\n\u001b[0m\u001b[1;32m     54\u001b[0m                                         inputs, attrs, num_outputs)\n\u001b[1;32m     55\u001b[0m   \u001b[0;32mexcept\u001b[0m \u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_NotOkStatusException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mInvalidArgumentError\u001b[0m: Graph execution error:\n\nDetected at node sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits defined at (most recent call last):\n  File \"/usr/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n\n  File \"/usr/lib/python3.10/runpy.py\", line 86, in _run_code\n\n  File \"/usr/local/lib/python3.10/dist-packages/colab_kernel_launcher.py\", line 37, in <module>\n\n  File \"/usr/local/lib/python3.10/dist-packages/traitlets/config/application.py\", line 992, in launch_instance\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelapp.py\", line 619, in start\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/platform/asyncio.py\", line 195, in start\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n\n  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 685, in <lambda>\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 738, in _run_callback\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 825, in inner\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 786, in run\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 361, in process_one\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 261, in dispatch_shell\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 539, in execute_request\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py\", line 302, in do_execute\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/zmqshell.py\", line 539, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2975, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3030, in _run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3257, in run_cell_async\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n\n  File \"<ipython-input-22-ddbb2e7932da>\", line 4, in <cell line: 4>\n\n  File \"<ipython-input-14-e97dc6108395>\", line 4, in train_autoencoder_with_early_stopping\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 65, in error_handler\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1807, in fit\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1401, in train_function\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1384, in step_function\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1373, in run_step\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1151, in train_step\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py\", line 1209, in compute_loss\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/engine/compile_utils.py\", line 277, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 143, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 270, in call\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses.py\", line 2454, in sparse_categorical_crossentropy\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend.py\", line 5775, in sparse_categorical_crossentropy\n\nReceived a label value of 128 which is outside the valid range of [0, 1).  Label values: 0 128 0 128 128 0 0 0 128 0 128 128 0 128 128 0 128 0 0 128 128 128 128 0 0 128 0 0 128 128 0 0\n\t [[{{node sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits}}]] [Op:__inference_train_function_838]"]}]}],"metadata":{"accelerator":"GPU","colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}